src/cis2/types.rs

```diff
@@ -0,0 +1,913 @@
+    #[display(
+        fmt = "{} {} as operator for {}",
+        update,
+        "display_address(operator)",
+        "display_address(owner)"
+    )]
+    UpdateOperator {
+        update:   OperatorUpdate,
+        owner:    Address,
+        operator: Address,
+    },
+    /// Provide an URL with the metadata for a certain token.
+    #[display(
+        fmt = "Added metadata url {} ({}) for token with ID {}",
+        "metadata_url.url",
+        "display_hash(metadata_url.hash)",
+        token_id
+    )]
+    TokenMetadata {
+        token_id:     TokenId,
+        metadata_url: MetadataUrl,
+    },
+    /// Custom event outside of the CIS2 specification.
+    #[display(fmt = "Unknown event: Event is not part of CIS2 specification")]
+    Unknown,
+}
+
+fn display_hash(hash_opt: Option<Hash>) -> String {
+    if let Some(hash) = hash_opt {
+        format!("with hash {}", hash)
+    } else {
+        "without hash".to_string()
+    }
+}
+
+/// Deserialize the contract events as according to the CIS2 specification.
+impl Deserial for Event {
+    fn deserial<R: Read>(source: &mut R) -> Result<Self, ParseError> {
+        let discriminant = u8::deserial(source)?;
+        match discriminant {
+            0 => Ok(Event::Transfer {
+                token_id: TokenId::deserial(source)?,
+                amount:   TokenAmount::deserial(source)?,
+                from:     Address::deserial(source)?,
+                to:       Address::deserial(source)?,
+            }),
+            1 => Ok(Event::Mint {
+                token_id: TokenId::deserial(source)?,
+                amount:   TokenAmount::deserial(source)?,
+                owner:    Address::deserial(source)?,
+            }),
+            2 => Ok(Event::Burn {
+                token_id: TokenId::deserial(source)?,
+                amount:   TokenAmount::deserial(source)?,
+                owner:    Address::deserial(source)?,
+            }),
+            3 => Ok(Event::UpdateOperator {
+                update:   OperatorUpdate::deserial(source)?,
+                owner:    Address::deserial(source)?,
+                operator: Address::deserial(source)?,
+            }),
+            4 => Ok(Event::TokenMetadata {
+                token_id:     TokenId::deserial(source)?,
+                metadata_url: MetadataUrl::deserial(source)?,
+            }),
+            _ => Ok(Event::Unknown),
+        }
+    }
+}
+
+/// Display the Address using either the display for account address or contract
+/// address.
+fn display_address(a: &Address) -> String {
+    match a {
+        Address::Account(addr) => format!("{}", addr),
+        Address::Contract(addr) => format!("{}", addr),
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::types::smart_contracts::concordium_contracts_common::{from_bytes, to_bytes};
+
+    #[test]
+    fn test_serialize_token_amount() {
+        for n in (0..1000).map(|i: u64| i.pow(3)) {
```

Comment:
THis is needlessly arbitrary. Either generate random values or just use a fixed range. Using powers of 3 is just strange.

Reply:
The reason was to ensure the coverage is spread enough.
Using random u64 will result in fewer small numbers.
Anyway, I am now using seeded random values and I have added some more unit tests with smaller values and cornercases.